Symbol uses

rojepp opened the pull request:
Implements #31.

I botched the previous PR by rebasing, so git wouldn't let me push (when will I ever learn).
I added a few fields from `SymbolUse`, including `IsFromDefinition` which should address @7sharp9's concern. Anything else that should be included?

rojepp committed:
Add symboluse command

rojepp committed:
Use proper types 

rojepp committed:
Add potentially useful fields to symboluse

7sharp9 commented:
I want to click the `Merge Pull Request` but Ill leave that to @rneatherway :-)

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -617,6 +641,36 @@ module internal Main =
+                              yield { StartLine = su.RangeAlternate.StartLine
+                                      StartColumn = su.RangeAlternate.StartColumn + 1
+                                      EndLine = su.RangeAlternate.EndLine
+                                      EndColumn = su.RangeAlternate.EndColumn + 1
```
Adding 1 here doesn't seem right, they should already be 1-based from the `Alternate` syntax. Looking at the results I also think they should end one earlier e.g. with:

```
let testval = FileTwo.NewObjectType()
```

The results have a range with columns start=5 and end=12. `t` is definitely at 5, and `l` is 6 further on, so 11. This is an inclusive range and I think we should leave it alone, documenting it as such (if we ever get around to adding documentation for this project!). If editors prefer to work with an exclusive range they can make the adjustement locally.

7sharp9 replied:
XS columns are 1 based we do something like:

```
        let start, finish = Symbols.trimSymbolRegion symbolUse lastIdentAtLoc
        let filename = doc.FileName.ToString()  
        let offset = doc.LocationToOffset(start.Line, start.Column+1)
        let domRegion = DomRegion(filename, start.Line, start.Column+1, finish.Line, finish.Column+1)

        let symbol = createSymbol(doc, context, symbolUse.Symbol, lastIdentAtLoc, domRegion)
        let memberRef = MemberReference(symbol, filename, offset, lastIdentAtLoc.Length)

        //if the current range is a symbol range and the fileNameOfRefs match change the ReferenceUsageType
        if symbolUse.IsFromDefinition then
            memberRef.ReferenceUsageType <- ReferenceUsageType.Write

        memberRef
```

rojepp replied:
I'm not sure how to fix this. FCS seems to report 0-based columns ([lines are 1-based](https://github.com/fsharp/FSharp.Compiler.Service/blob/251c00058cb4ec48e9fc31a7a814d8a3f4898d3f/src/fsharp/vs/service.fsi#L136)). Do you want me to only add 1 to `StartColumn`? That doesn't seem right to me. That would leave `testva` highlighted.

rneatherway replied:
Sorry, I misread. It seems that FCS [accepts 1-based column numbers](https://github.com/fsharp/FSharp.Compiler.Service/blob/251c00058cb4ec48e9fc31a7a814d8a3f4898d3f/src/fsharp/vs/service.fsi#L189) but [returns 0-based](https://github.com/fsharp/FsAutoComplete/blob/master/FSharp.AutoComplete/test/integration/FindDeclarations/output.txt#L18). Why would this possibly be a good idea?

I've looked at the output we give and I think only `declarations`, `finddecl` and this new command return positions. I think they should all return 0-based or all return 1-based columns, for consistency. However, I think that if FCS is outputting 0-based columns it should probably be accepting 0-based as well.

rojepp replied:
It is all very confusing. Adding 1 to `StartColumn` and `EndColumn` seems to be the right thing to do, so we normalize to consistently being 1-based . Do the other commands return 0-based?

rneatherway replied:
I agree with you that it is very confusing. If you look at the test output I linked in the last comment I found that `GetDeclaration` returns 0-based. So does the one that returns all the declarations in the file, but that isn't really used at the moment.

rojepp replied:
Find Declaration seems to return 0-based.

rneatherway replied:
Yes, I guess you posted at the same time as me ;-)

I would be happy if everything returned 1-based so at least FSAC is consistent even if FCS isn't.

@7sharp9 do you adjust the results you get from `GetNavigationItems` for the pathed document support as well to be 1-based?

rojepp replied:
Let's do it. :) 

Separate PR or this one?

rneatherway replied:
Here is fine by me.

rojepp replied:
Done.

rneatherway commented:
Otherwise looks great. Thanks so much for the contribution, things are really coming on!

rojepp committed:
Make ouput columns 1-based

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -461,7 +485,7 @@ module internal Main =
+                  sb.AppendLine(sprintf "[%d:%d-%d:%d] %s %s" e.StartLineAlternate (e.StartColumn + 1) e.EndLineAlternate (e.EndColumn + 1)
                                   (if e.Severity = FSharpErrorSeverity.Error then "ERROR" else "WARNING") e.Message)
                   |> ignore
                 sb.Append("<<EOF>>") |> ignore
```
The JSON printing just below `Data =errs` is also affected

rojepp replied:
Good catch, I'll fix it shortly.

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -521,11 +545,11 @@ module internal Main =
+                      let (s1, e1), (s2, e2) = ((m.StartColumn + 1, m.StartLine), (m.EndColumn + 1, m.EndLine))
                       yield sprintf "  - [%d:%d-%d:%d] %s" e1 s1 e2 s2 d.Name ]
               printAgent.WriteLine(sprintf "DATA: declarations\n%s\n<<EOF>>" (String.concat "\n" declstrings))
           | Json -> prAsJson { Kind = "declarations"; Data = decls }
```
This line is also affected

rojepp replied:
This one could be handled in the Range converter? 
https://github.com/fsharp/FsAutoComplete/blob/master/FSharp.AutoComplete/Program.fs#L111

rojepp replied:
OTOH, maybe we should have a record type for this as well. It'd be nice to have StartColumn/StartLine etc in this type as well, for consistency.

rneatherway commented:
Having to also correct the errors makes it less enjoyable :-(

rojepp commented:
Argh, `FSharpErrorInfo` has no public constructor. I guess the only option is to make a record type with the same fields.

rojepp commented:
In doing a record type, should I keep `StartLineAlternate`/`EndLineAlternate`? Maybe a good time to expose fewer fields?

rojepp committed:
Make output columns 1-based for FSharpErrorInfo

In doing so, I also removed FSharpErrorInfo.StartLineAlternate and FSharpErrorInfo.EndLineAlternate. Now there is only StartLine/EndLine, and they're 1-based.

rojepp committed:
Serialize Ranges as 1-based

rojepp commented:
With that, I _think_ I'm done. :)

rojepp committed:
Remove unused var warning
